File: frequency-tracker-2671/Solution.java

/**
 * LeetCode problem # 2671, 'Frequency Tracker'
 * https://leetcode.com/problems/frequency-tracker
 */

import java.util.*;

class FrequencyTracker {
    Map<Integer, Integer> map;
    int[] freq;

    public FrequencyTracker() {
        map = new HashMap<>();
        freq = new int[100001];
    }

    public void add(int number) {
        int curr = map.getOrDefault(number, 0);
        int next = curr + 1;
        map.put(number, next);
        freq[curr]--;
        freq[next]++;
    }

    public void deleteOne(int number) {
        if (map.containsKey(number)) {
            int curr = map.get(number);
            int next = curr - 1;
            if (next == 0) {
                map.remove(number);
            } else {
                map.put(number, next);
            }
            freq[curr]--;
            freq[next]++;
        }
    }

    public boolean hasFrequency(int frequency) {
        return freq[frequency] > 0;
    }
}

/**
 * Your FrequencyTracker object will be instantiated and called as such:
 * FrequencyTracker obj = new FrequencyTracker(); obj.add(number);
 * obj.deleteOne(number); boolean param_3 = obj.hasFrequency(frequency);
 */

File: frequency-tracker-2671/solution.js
/**
 * LeetCode problem # 2671, 'Frequency Tracker'
 * https://leetcode.com/problems/frequency-tracker
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
function findSecondMinimumValue(root) {
  if (!root) {
    return -1;
  }

  let min = Infinity;
  let prevMin = -1;

  const traverse = (node) => {
    if (!node) {
      return;
    }
    if (node.val < min) {
      prevMin = min;
      min = node.val;
    } else if (node.val > min) {
      prevMin = prevMin < node.val ? prevMin : node.val;
    }
    traverse(node.left);
    traverse(node.right);
  };

  traverse(root);
  return prevMin == Infinity ? -1 : prevMin;
}

/**
 * Your FrequencyTracker object will be instantiated and called as such:
 * FrequencyTracker obj = new FrequencyTracker(); obj.add(number);
 * obj.deleteOne(number); boolean param_3 = obj.hasFrequency(frequency);
 */
